preprocessing.py

```diff
@@ -0,0 +1,182 @@
+import json
+import os
+import argparse
+import cv2
+import numpy as np
+import open3d
+import configparser
+import imageio
+
+
+def undistort(path, camera_matrix, dist_coeff):
+    """
+    Undistorts image given
+    :param path: path to the image for undistorting
+    :param camera_matrix: camera's intrinsic parameters
+    :param dist_coeff: list of k1, k2, p1, p2, k3, k4, k5, k6 coefficients
+    :return: undistorted image and new camera's intrinsics
+    """
+    image = cv2.imread(path, -1)
+    shape = image.shape[:2][::-1]
+    undist_intrinsics, _ = cv2.getOptimalNewCameraMatrix(
+        camera_matrix, dist_coeff, shape, 1, shape
+    )
+    map_x, map_y = cv2.initUndistortRectifyMap(
+        camera_matrix, dist_coeff, None, undist_intrinsics, shape, cv2.CV_32FC1
+    )
+    undistorted = cv2.remap(image, map_x, map_y, cv2.INTER_NEAREST)
+    return undistorted, undist_intrinsics
+
+
+def reprojection(
+    depth_undistorted,
+    undist_intrinsics,
+    intrinsic,
+    transform_matrix,
+    color_matrix,
+    shape,
+):
+    """
+    Reprojects the points of the depth image onto the coordinates of the color image
+    :param depth_undistorted: undistorted depth image
+    :param undist_intrinsics: camera's intrinsics after undistort
+    :param intrinsic: open3d.camera.PinholeCameraIntrinsic instance
+    :param transform_matrix: transformation matrix from one coordinate system to another
+    :param color_matrix: color camera's intrinsics
+    :param shape: shape of color images
+    :return: new depth image in the color image coordinates
+    """
+    picture = open3d.geometry.Image(depth_undistorted)
+    intrinsic.intrinsic_matrix = undist_intrinsics
+    pc = open3d.geometry.PointCloud()
+    point_cloud = pc.create_from_depth_image(picture, intrinsic)
+
+    point_cloud = point_cloud.transform(transform_matrix)
+    points = np.concatenate(
+        (np.asarray(point_cloud.points), np.full((len(point_cloud.points), 1), 1)),
+        axis=1,
+    )
+
+    result = color_matrix @ points.T
+    result[0] /= result[2]
+    result[1] /= result[2]
+    result[2] *= 1000
+    result = np.round(result).astype(int)
+    w, h = shape
+    img = np.zeros((h, w)).astype(np.uint16)
+    for point in result.T:
+        u, v, z, _ = point
+        if 0 <= v < h and 0 <= u < w:
```

Comment:
You can filter out points from `result` using conditions (loops are terrible from performance POV, try to avoid them):
Pseudocode:
```
a = a[:, (a[0] >= 0) & (a[0] < w)]
```